State sync support

erikgrinaker opened the pull request:
## Description

Fixes #5689, fixes #5690. Adds support for taking, restoring, and serving snapshots for [state sync](https://github.com/tendermint/tendermint/issues/828), as outlined in [ADR-053](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-053-state-sync-prototype.md) and the [ABCI reference](https://docs.tendermint.com/master/spec/abci/apps.html#state-sync).

* `rootmulti.Store` now implements a new `snapshots.Snapshotter` interface, which can take and restore binary state snapshots at a given height. Snapshots are built as follows:

  1. `rootmulti.Store` iterates over all `iavl.Store` stores, and for each one emits a delimited Protobuf `SnapshotStoreItem` message with store metadata

  2. For each `iavl.Store`, export `iavl.ExportNode` nodes via `Export()` at the given height and emit a delimited Protobuf `SnapshotIAVLItem` message for each node

  3. The delimited Protobuf message stream is zlib-compressed

  4. The zlib-compressed stream is chunked naïvely into 10 MB chunks

* `snapshots.Store` stores snapshot metadata in a separate database and chunks in a filesystem directory

* `BaseApp` takes asynchronous snapshots after `Commit()` in regular height intervals given by the config option `state-sync.snapshot-interval` (default 0, i.e. disabled). It also prunes old snapshots, and keeps the most recent snapshots given by the option `state-sync.snapshot-keep-recent` (default 2).

* `BaseApp` implements the upcoming state sync ABCI interface from Tendermint, for fetching and applying state snapshots.

* `SimApp` has been updated to enable and initialize state sync snapshots.

alexanderbez commented on baseapp/abci.go:
```diff
@@ -262,6 +262,27 @@ func (app *BaseApp) Commit() (res abci.ResponseCommit) {
+			app.logger.Info("Finished taking state snapshot", "height", header.Height)
+			if app.snapshotRetention > 0 {
+				app.logger.Info("Pruning state snapshots")
+				pruned, err := app.snapshotStore.Prune(app.snapshotRetention)
```
we should probably have an auxiliary method to prune just like we do for snapshotting that does preliminary checks and such.

erikgrinaker replied:
Yeah, that makes sense.

alexanderbez commented on server/constructors.go:
```diff
@@ -11,13 +11,14 @@ import (
 	// AppCreator is a function that allows us to lazily initialize an
 	// application using various configurations.
-	AppCreator func(log.Logger, dbm.DB, io.Writer) abci.Application
+	AppCreator func(log.Logger, dbm.DB, *snapshot.Store, io.Writer) abci.Application
```
The snapshot store along with the policy will be passed as options to the baseapp in the app's constructor (correct me if I'm wrong). Why not just have the constructor itself create the snapshot store if it wishes instead of modying the `AppCreator`?

erikgrinaker replied:
I'm cleaning this up a bit now. I would really like to just default to a `snapshots` directory under `BaseConfig.RootDir`. The only place where we have access to `RootDir` is e.g. here: https://github.com/cosmos/cosmos-sdk/blob/6eb2ffed55010f65d303c0106e0e26271d0b3ec1/server/start.go#L180

Do you have any suggestions for how to pass this to the application automatically without modifying `AppCreator`?

erikgrinaker replied:
Nevermind, decided to handle this in Gaia instead, placing snapshots under `gaiad/snapshots/`.

alexanderbez commented on store/rootmulti/store.go:
```diff
@@ -485,6 +496,195 @@ func parsePath(path string) (storeName string, subpath string, err error) {
+		bufWriter := bufio.NewWriterSize(chunkWriter, snapshotBufferSize)
+		defer func() {
+			if err := bufWriter.Flush(); err != nil {
+				chunkWriter.CloseWithError(err)
```
will this cause an issue with the previous `defer`s `Close`?

alexanderbez commented on store/snapshot/store.go:
```diff
@@ -0,0 +1,354 @@
+package snapshot
```
I agree that this probably doesn't belong under `store/`. I would reason that I should probably be its own root package (it's not a module either). In addition, I'd also move the snapshot types you have in `types/` to this new root package as well.

erikgrinaker replied:
>I'd also move the snapshot types you have in types/ to this new root package as well.

Well, it depends on how we structure this. I was thinking the snapshot store shouldn't care about what the snapshots actually contain, they're just opaque binary blobs. The actual snapshotting is the responsibility of the data store (i.e. `rootmulti.Store`). This decouples them cleanly, such that the snapshot store doesn't have to know about the data store's internals. Thus the `SnapshotItem` and related types (snapshot contents) go with `rootmulti.Store`, while `SnapshotMetadata` and `SnapshotChunkMetadata` go with `snapshot.Store`.

alexanderbez replied:
Makes sense! But I'd still argue that they don't belong in `types/`. If they are specific to how the root multi-store behaves, shouldn't they be defined in scope of the root multi-store (i.e. the store pkg).

erikgrinaker replied:
Yes, you're right of course. I added them in the root types because I didn't see any Protobufs for the `store` types, and forgot about it. I'll reorganize it a bit.

alexanderbez commented on store/rootmulti/utils.go:
```diff
@@ -0,0 +1,144 @@
+package rootmulti
```
I have a hunch these should live under the snapshot package, thoughts?

erikgrinaker replied:
See comment above. These utility methods are used for snapshot contents, not the metadata, so they belong with whatever does the actual data store snapshotting.

erikgrinaker commented:
We've been discussing how we should configure block retention and snapshotting, there's an RFC on this here: https://github.com/tendermint/spec/pull/84

I suggest we hold off on merging this until the RFC is resolved.

erikgrinaker commented:
We'll have to come up with a different snapshot format or chunking scheme: neither Protobuf nor gzip is stable/deterministic across environments, thus chunks will not be identical across nodes. I'll give it some thought over the weekend.

alexanderbez commented:
> We'll have to come up with a different snapshot format or chunking scheme: neither Protobuf nor gzip is stable/deterministic across environments, thus chunks will not be identical across nodes. I'll give it some thought over the weekend.

@erikgrinaker can you elaborate on Protobuf not being stable or deterministic across environments? Proto3 should be deterministic across binaries. Where do you see the determinism breaking down?

erikgrinaker commented:
> can you elaborate on Protobuf not being stable or deterministic across environments

A few examples would be that the spec doesn't require a stable/deterministic order e.g. of struct fields, fields may be omitted, and varints are not required to be optimally packed: https://gist.github.com/kchristidis/39c8b310fd9da43d515c4394c3cd9510

Thus if we e.g. update the protobuf library or generate code using a different version of `protoc` this may cause the chunks to be generated differently. The current chunker requires chunks to be identical byte-by-byte across nodes.

alexanderbez commented:
Right, but it should respect field ordering, no? I know certain clients may omit fields and others don't, but ordering should be enforced IIRC. I could be wrong and if so, this would be problematic.

I would imagine if we use Protobuf in the state-machine (which we do), it should be good enough for chunking the state of the state-machine.

erikgrinaker commented:
The difference is that the state machine only cares about the semantic contents of the protobufs, while the chunker cares about the serialized byte layout. Let's consider two nodes A and B where one omits empty fields and the other doesn't:

| Node | Message |
|------|--------|
| A | `1=foo 2= 3=barbaz` |
| B | `1=foo 3=barbaz ` |

Let's say the chunker splits after byte 12. The chunks now look like this:

| Node | Chunk 1 | Chunk 2 |
|------|--------|-------|
| A | `1=foo 2= 3=b` | `arbaz` |
| B | `1=foo 3=barb` | `az ` |

If we pick chunk 1 from node A and chunk 2 from node B, the received message is `1=foo 2=  3=baz`. This is incorrect. The same problem applies to e.g. field ordering and other aspects as well.

There are a few obvious options here, e.g.:

* Have tests that make sure a given dataset always serializes to the same byte stream.
* Use logical chunks instead, where each chunk can be decoded individually.
* Use a deterministic serializer.

I'll give this some thought over the weekend. I think the compression algorithm is a bigger problem than Protobuf (gzip in particular, zlib or deflate may be better).

alexanderbez commented:
Thanks for the clarification @erikgrinaker! Note I still don't see how this is different from the state-machine. We write the proto3 serialized bytes which are constituents of the Merkle root.

erikgrinaker commented:
Ah, I suppose that's true. Then yes, the state machine has the same problem - if different versions of the state machine with different protobuf serializers execute a given transaction then they can produce non-deterministic results, and only a state machine with a +2/3 majority will be able to make progress.

erikgrinaker commented:
I've changed snapshots to use `zlib` compression (which is more appropriate anyway), and added a checksum test for snapshot byte stream stability. As long as we can reasonably expect all state machines across a network to run the same binary or version, I guess this should be good enough for now. If necessary we can improve this later without breaking any APIs. Thanks for the input @alexanderbez!

alexanderbez commented:
Great! Good idea on the streamed checksum 👍

marbar3778 commented:
leave it open

erikgrinaker commented:
I've ported this to the latest SDK+Tendermint target in the `marko/tm_update` branch as `erik/state-sync`.

erikgrinaker commented:
Superseded by #7166.
